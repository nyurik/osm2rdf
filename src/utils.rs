use chrono::{TimeZone, Utc};
use json::JsonValue;
use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
use regex::Regex;

pub(crate) trait StringExt {
    fn push_str_value(&mut self, predicate: &'static str, value: &str);
    fn push_quoted_str(&mut self, value: &str);
    fn push_char_value(&mut self, predicate: &'static str, value: char);
    fn push_bool_value(&mut self, predicate: &'static str, value: bool);
    fn push_int_value(&mut self, predicate: &'static str, value: i64);
    fn push_date_value(&mut self, predicate: &'static str, value: i64);
    fn push_point(&mut self, predicate: &'static str, latitude: f64, longitude: f64);
    fn push_wiki_url(&mut self, lang: &str, site: &str, title: &str);
    fn push_tag(&mut self, key: &str, value: &str, utils: &Consts);
    fn push_metadata(&mut self, version: i32, user: &str, milli_timestamp: i64, changeset: i64);
}

impl StringExt for String {
    fn push_str_value(&mut self, predicate: &'static str, value: &str) {
        self.push_str(predicate);
        self.push(' ');
        self.push_quoted_str(value);
    }

    fn push_quoted_str(&mut self, value: &str) {
        // TODO: optimize?
        self.push_str(&JsonValue::from(value).dump());
        self.push_str(";\n");
    }

    fn push_char_value(&mut self, predicate: &'static str, value: char) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push(value);
        self.push_str("\";\n");
    }

    fn push_bool_value(&mut self, predicate: &'static str, value: bool) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(if value { "true" } else { "false" });
        self.push_str("\"^^xsd:boolean;\n");
    }

    fn push_int_value(&mut self, predicate: &'static str, value: i64) {
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(&value.to_string());
        self.push_str("\"^^xsd:integer;\n");
    }

    fn push_date_value(&mut self, predicate: &'static str, milli_timestamp: i64) {
        // "{0:%Y-%m-%dT%H:%M:%S}Z"^^xsd:dateTime
        self.push_str(predicate);
        self.push_str(" \"");
        self.push_str(&format_ts(milli_timestamp));
        self.push_str("\"^^xsd:dateTime;\n");
    }

    fn push_point(&mut self, predicate: &'static str, latitude: f64, longitude: f64) {
        self.push_str(predicate);
        self.push_str(" \"Point(");
        self.push_str(&longitude.to_string());
        self.push(' ');
        self.push_str(&latitude.to_string());
        self.push_str(")\"^^geo:wktLiteral");
        self.push_str(";\n");
    }

    fn push_wiki_url(&mut self, lang: &str, site: &str, title: &str) {
        self.push_str("<https://");
        self.push_str(lang);
        self.push_str(site);
        self.extend(utf8_percent_encode(
            title.replace(' ', "_").as_str(),
            PERCENT_ENC_SET,
        ));
        self.push('>');
        self.push_str(";\n");
    }

    fn push_tag(&mut self, key: &str, value: &str, consts: &Consts) {
        if !consts.re_simple_local_name.is_match(key) {
            // Record any unusual tag name in a "osmm:badkey" statement
            self.push_str_value("osmm:badkey", value);
            return;
        }

        self.push_str("osmt:");
        self.push_str(key);
        self.push(' ');
        let mut parsed = false;
        if key.contains("wikidata") {
            if consts.re_wikidata_value.is_match(value) {
                self.push_str("wd:");
                self.push_str(value);
                parsed = true;
            } else if consts.re_wikidata_multi_value.is_match(value) {
                for v in value.split(';') {
                    self.push_str("wd:");
                    self.push_str(v);
                    self.push(',');
                }
                self.pop(); // remove trailing ","
                parsed = true;
            }
        } else if key.contains("wikipedia") {
            if let Some(v) = consts.re_wikipedia_value.captures(value) {
                self.push_wiki_url(
                    v.get(1).unwrap().as_str(),
                    ".wikipedia.org/wiki/",
                    v.get(2).unwrap().as_str(),
                );
                parsed = true;
            }
        }
        if !parsed {
            self.push_quoted_str(value);
        }
    }

    fn push_metadata(&mut self, version: i32, user: &str, milli_timestamp: i64, changeset: i64) {
        self.push_int_value("osmm:version", version as i64);
        self.push_str_value("osmm:user", user);
        self.push_date_value("osmm:timestamp", milli_timestamp);
        self.push_int_value("osmm:changeset", changeset);
        self.pop();
        self.pop();
        self.push_str(".\n");
    }
}

pub fn format_ts(milli_timestamp: i64) -> String {
    format!(
        "{:?}",
        Utc.timestamp(milli_timestamp / 1000, (milli_timestamp % 1000) as u32)
    )
}

pub const PERCENT_ENC_SET: &AsciiSet = &CONTROLS
    .add(b';')
    .add(b'@')
    .add(b'$')
    .add(b'!')
    .add(b'*')
    .add(b'(')
    .add(b')')
    .add(b',')
    .add(b'/')
    .add(b'~')
    .add(b':')
    // The "#" is also safe - used for anchoring
    .add(b'#');

#[derive(Clone, Debug)]
pub struct Consts {
    pub re_simple_local_name: Regex,
    pub re_wikidata_key: Regex,
    pub re_wikidata_value: Regex,
    pub re_wikidata_multi_value: Regex,
    pub re_wikipedia_value: Regex,
}

impl Consts {
    pub fn new() -> Self {
        Consts {
            /// Total length of the maximum "valid" local name is 60 (58 + first + last char)
            /// Local name may contain letters, numbers anywhere, and -:_ symbols anywhere except first and last position
            re_simple_local_name: Regex::new(r"^[0-9a-zA-Z_]([-:0-9a-zA-Z_]{0,58}[0-9a-zA-Z_])?$")
                .unwrap(),

            re_wikidata_key: Regex::new(r"(.:)?wikidata$").unwrap(),
            re_wikidata_value: Regex::new(r"^Q[1-9][0-9]{0,18}$").unwrap(),
            re_wikidata_multi_value: Regex::new(r"^Q[1-9][0-9]{0,18}(;Q[1-9][0-9]{0,18})+$")
                .unwrap(),
            re_wikipedia_value: Regex::new(r"^([-a-z]+):(.+)$").unwrap(),
        }
    }
}

#[derive(Clone, Default, Debug)]
pub struct Stats {
    pub added_nodes: u64,
    pub added_rels: u64,
    pub added_ways: u64,
    pub skipped_nodes: u64,
    pub deleted_nodes: u64,
    pub deleted_rels: u64,
    pub deleted_ways: u64,
    pub blocks: u64,
}

impl Stats {
    pub(crate) fn combine(&mut self, other: Stats) {
        self.added_nodes += other.added_nodes;
        self.added_rels += other.added_rels;
        self.added_ways += other.added_ways;
        self.skipped_nodes += other.skipped_nodes;
        self.deleted_nodes += other.deleted_nodes;
        self.deleted_rels += other.deleted_rels;
        self.deleted_ways += other.deleted_ways;
        self.blocks += 1;
    }
}

#[derive(Debug)]
pub enum Statement {
    Skip,
    Delete {
        elem: Element,
        id: i64,
    },
    Create {
        elem: Element,
        id: i64,
        ts: i64,
        val: String,
    },
}

#[derive(Debug)]
pub enum Element {
    Node,
    Way,
    Relation,
}
