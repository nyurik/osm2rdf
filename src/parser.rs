use geos::{CoordSeq, Geom, Geometry};
use osmnodecache::{Cache, DenseFileCache, DenseFileCacheOpts};
use osmpbf::{BlobDecode, BlobReader, DenseNode, Info, Node, RelMemberType, Relation, Way};

use crate::utils::{format_ts, Consts, Element, Statement, Stats, StringExt};
use crate::{Command, Opt};
use anyhow::Error;
use byteorder::WriteBytesExt;
use bytesize::ByteSize;
use flate2::write::GzEncoder;
use flate2::Compression;
use path_absolutize::Absolutize;
use rayon::iter::{ParallelBridge, ParallelIterator};
use std::fs::File;
use std::io::Write;
use std::path::PathBuf;
use std::sync::atomic::{AtomicI64, AtomicU32, Ordering};
use std::sync::mpsc::{channel, Receiver};
use std::sync::Mutex;
use std::thread::{Builder, JoinHandle};

struct Parser<'a> {
    parent_stats: &'a Mutex<Stats>,
    stats: Stats,
    consts: Consts,
    cache: Box<dyn Cache + 'a>,
}

impl<'a> Drop for Parser<'a> {
    fn drop(&mut self) {
        let stats = std::mem::take(&mut self.stats);
        self.parent_stats.lock().unwrap().combine(stats);
    }
}

impl<'a> Parser<'a> {
    fn new(
        parent_stats: &'a Mutex<Stats>,
        consts: &Consts,
        dfc: &'a mut DenseFileCache,
    ) -> Parser<'a> {
        Parser {
            parent_stats,
            stats: Stats::default(),
            consts: consts.clone(),
            cache: Box::new(dfc.get_accessor()),
        }
    }

    fn on_node(&mut self, node: &Node) -> Statement {
        self.cache.set_value_i32(
            node.id() as usize,
            node.decimicro_lat(),
            node.decimicro_lon(),
        );
        let info = &node.info();
        let mut statement = self.process_node(
            node.info().deleted(),
            node.id(),
            node.tags(),
            node.lat(),
            node.lon(),
        );
        if let Statement::Create {
            val: ref mut value,
            ref mut ts,
            ..
        } = statement
        {
            *ts = Self::push_info(value, info);
        }
        statement
    }

    fn on_dense_node(&mut self, node: &DenseNode) -> Statement {
        self.cache.set_value_i32(
            node.id() as usize,
            node.decimicro_lat(),
            node.decimicro_lon(),
        );
        let info = node.info().unwrap();
        let mut statement = self.process_node(
            info.deleted(),
            node.id(),
            node.tags(),
            node.lat(),
            node.lon(),
        );
        if let Statement::Create {
            val: ref mut value, ..
        } = statement
        {
            value.push_metadata(
                info.version(),
                info.user().unwrap(),
                info.milli_timestamp(),
                info.changeset(),
            );
        }
        statement
    }

    fn process_node<'t, TTags: Iterator<Item = (&'t str, &'t str)> + ExactSizeIterator>(
        &mut self,
        is_deleted: bool,
        id: i64,
        tags: TTags,
        lat: f64,
        lon: f64,
    ) -> Statement {
        if is_deleted {
            self.stats.deleted_nodes += 1;
            Statement::Delete {
                elem: Element::Node,
                id,
            }
        } else {
            let mut value = String::with_capacity(100000);
            self.push_all_tags(&mut value, tags);
            if value.is_empty() {
                self.stats.skipped_nodes += 1;
                Statement::Skip
            } else {
                value.push_point("osmm:loc", lat, lon);
                Self::push_elem_type(&mut value, Element::Node);
                self.stats.added_nodes += 1;
                Statement::Create {
                    elem: Element::Node,
                    id,
                    val: value,
                    ts: 0,
                }
            }
        }
    }

    fn on_way(&mut self, way: &Way) -> Statement {
        let info = way.info();
        if info.deleted() {
            self.stats.deleted_ways += 1;
            return Statement::Delete {
                elem: Element::Way,
                id: way.id(),
            };
        }
        let mut value = String::with_capacity(100000);
        self.push_all_tags(&mut value, way.tags());
        Self::push_elem_type(&mut value, Element::Way);
        let ts = Self::push_info(&mut value, &info);
        if let Err(err) = self.parse_way_geometry(&mut value, way) {
            value.push_tag("osmm:loc:error", &err.to_string(), &self.consts)
        }

        self.stats.added_ways += 1;
        Statement::Create {
            elem: Element::Way,
            id: way.id(),
            val: value,
            ts,
        }
    }

    fn on_relation(&mut self, rel: &Relation) -> Statement {
        let info = rel.info();
        if info.deleted() {
            self.stats.deleted_rels += 1;
            return Statement::Delete {
                elem: Element::Way,
                id: rel.id(),
            };
        }
        let mut value = String::with_capacity(100000);
        self.push_all_tags(&mut value, rel.tags());
        Self::push_elem_type(&mut value, Element::Relation);
        let ts = Self::push_info(&mut value, &info);
        for mbr in rel.members() {
            // Produce two statements - one to find all members of a relation,
            // and another to find the role of that relation
            //     osmrel:123  osmm:has    osmway:456
            //     osmrel:123  osmway:456  "inner"
            let mbr_prefix = match mbr.member_type {
                RelMemberType::Node => "osmnode:",
                RelMemberType::Way => "osmway:",
                RelMemberType::Relation => "osmrel:",
            };
            let mbr_id = mbr.member_id.to_string();

            value.push_str("osmm:has ");
            value.push_str(mbr_prefix);
            value.push_str(&mbr_id);
            value.push_str(";\n");
            value.push_str(mbr_prefix);
            value.push_str(&mbr_id);
            value.push(' ');
            value.push_quoted_str(mbr.role().unwrap());
        }

        self.stats.added_rels += 1;
        Statement::Create {
            elem: Element::Relation,
            id: rel.id(),
            val: value,
            ts,
        }
    }

    fn push_info(value: &mut String, info: &Info) -> i64 {
        let ts = info.milli_timestamp().unwrap();
        value.push_metadata(
            info.version().unwrap(),
            info.user().unwrap().unwrap(),
            ts,
            info.changeset().unwrap(),
        );
        ts
    }

    fn push_elem_type(value: &mut String, element: Element) {
        value.push_char_value(
            "osmm:type",
            match element {
                Element::Node => 'n',
                Element::Way => 'w',
                Element::Relation => 'r',
            },
        );
    }

    fn parse_way_geometry(&self, value: &mut String, way: &Way) -> anyhow::Result<()> {
        let refs: Vec<[f64; 2]> = way
            .refs()
            .map(|id| {
                let (lat, lng) = self.cache.get_lat_long(id as usize);
                [lat as f64, lng as f64]
            })
            .collect();

        let geometry = Geometry::create_line_string(CoordSeq::new_from_vec(&refs)?)?;
        value.push_bool_value("osmm:isClosed", geometry.is_closed()?);
        let g = geometry.point_on_surface()?;
        value.push_point("osmm:loc", g.get_y().unwrap(), g.get_x().unwrap());

        Ok(())
    }

    fn push_all_tags<'t, TTags: Iterator<Item = (&'t str, &'t str)> + ExactSizeIterator>(
        &mut self,
        value: &mut String,
        tags: TTags,
    ) {
        for (k, v) in tags {
            if k != "created_by" {
                value.push_tag(k, v, &self.consts);
            }
        }
    }
}

fn create_cache(filename: PathBuf) -> anyhow::Result<DenseFileCache> {
    DenseFileCacheOpts::new(filename)
        .page_size(10 * 1024 * 1024 * 1024)
        .on_size_change(Some(|old_size, new_size| {
            println!(
                "Growing cache {} âž¡ {}",
                ByteSize(old_size as u64),
                ByteSize(new_size as u64)
            )
        }))
        .open()
}

fn start_writer_thread(
    output_dir: &PathBuf,
    max_file_size: usize,
    receiver: Receiver<Statement>,
) -> JoinHandle<()> {
    let output_dir = output_dir.clone();
    let file_index = AtomicU32::new(0);
    let oldest_ts = AtomicI64::new(0);

    Builder::new()
        .name("gz_writer".into())
        .spawn(move || {
            let mut encoder = None;
            let mut size = 0_usize;
            while let Ok(v) = receiver.recv() {
                if let Statement::Create { elem, id, val, ts } = v {
                    let enc = encoder.get_or_insert_with(|| new_gz_file(&output_dir, &file_index));

                    let prefix = match elem {
                        Element::Node => "osmnode:",
                        Element::Way => "osmway:",
                        Element::Relation => "osmrel:",
                    };
                    oldest_ts.fetch_max(ts, Ordering::Relaxed);

                    enc.write_all(prefix.as_ref()).unwrap();
                    enc.write_all(id.to_string().as_ref()).unwrap();
                    enc.write_u8(b'\n').unwrap();
                    enc.write_all(val.as_ref()).unwrap();
                    enc.write_u8(b'\n').unwrap();

                    size += val.len();
                    if size > max_file_size {
                        encoder.take().unwrap().finish().unwrap();
                        size = 0;
                    }
                }
            }

            let mut enc = new_gz_file(&output_dir, &file_index);
            let ts = format_ts(oldest_ts.load(Ordering::SeqCst));
            let statement = format!("osmroot: schema:dateModified {}.\n", ts);
            enc.write_all(statement.as_ref()).unwrap();
        })
        .unwrap()
}

fn new_gz_file(output_dir: &PathBuf, file_index: &AtomicU32) -> GzEncoder<File> {
    let index = file_index.fetch_add(1, Ordering::Relaxed);
    let file = output_dir.join(format!("osm-{:06}.ttl.gz", index));
    println!("Creating {:?}", file.absolutize().unwrap());
    GzEncoder::new(File::create(file).unwrap(), Compression::default())
}

pub fn parse(opt: Opt) -> Result<(), Error> {
    match opt.cmd {
        Command::Parse {
            workers,
            input_file,
            output_dir,
            max_file_size,
        } => {
            if let Some(v) = workers {
                rayon::ThreadPoolBuilder::new()
                    .thread_name(|i| format!("parser #{}", i))
                    .num_threads(v)
                    .build_global()
                    .unwrap();
            }
            let reader = BlobReader::from_path(input_file)?;
            let file_cache = create_cache(opt.cache)?;
            let stats = Mutex::new(Stats::default());
            let consts = &Consts::new();
            let (sender, receiver) = channel();
            let writer_thread =
                start_writer_thread(&output_dir, max_file_size * 1024 * 1024, receiver);

            reader
                .par_bridge()
                .for_each_with((file_cache, sender), |(dfc, writer), blob| {
                    let mut parser = Parser::new(&stats, consts, dfc);
                    if let BlobDecode::OsmData(block) = blob.unwrap().decode().unwrap() {
                        for group in block.groups() {
                            // FIXME: possible concurrency bug: a non-node element may need coords of a node that hasn't been processed yet
                            for node in group.nodes() {
                                writer.send(parser.on_node(&node)).unwrap();
                            }
                            for node in group.dense_nodes() {
                                writer.send(parser.on_dense_node(&node)).unwrap();
                            }
                            for way in group.ways() {
                                writer.send(parser.on_way(&way)).unwrap();
                            }
                            for rel in group.relations() {
                                writer.send(parser.on_relation(&rel)).unwrap();
                            }
                        }
                    };
                });

            writer_thread.join().unwrap();
            println!("Summary statistics:\n{:#?}", stats.lock().unwrap());
            Ok(())
        } // _ => panic!("Expecting Parse")
    }
}
