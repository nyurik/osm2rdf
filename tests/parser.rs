use insta::glob;
use osm2rdf::parser::parse_block;
use osm2rdf::parser::Parser;
use osm2rdf::utils::Stats;
use osmnodecache::{CacheStore, HashMapCache};
use osmpbf::{BlobDecode, BlobReader};
use std::sync::Mutex;

#[test]
fn decode_osm_pbf_files() {
    glob!("../tests/fixtures", "**/*.osm.pbf", |file| {
        let reader = BlobReader::from_path(file).unwrap();
        let cache = HashMapCache::new();
        let stats = Mutex::new(Stats::default());
        let mut parser = Parser::new(&stats, cache.get_accessor());

        let mut result = Vec::new();
        for blob in reader {
            if let BlobDecode::OsmData(block) = blob.unwrap().decode().unwrap() {
                parse_block(block, &mut parser, |s| result.push(s));
            };
        }
        insta::assert_debug_snapshot!(result);
    });
}
