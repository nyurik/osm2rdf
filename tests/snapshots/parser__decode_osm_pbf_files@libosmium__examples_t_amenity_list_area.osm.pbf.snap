---
source: tests/parser.rs
expression: result
input_file: tests/fixtures/libosmium/examples_t_amenity_list_area.osm.pbf
---
[
    Skip,
    Skip,
    Skip,
    Skip,
    Create {
        elem: Way,
        id: 274425514,
        ts: 1471461400000,
        val: indoc! {r#"
            osmt:amenity "restaurant";
            osmt:building "house";
            osmt:cuisine "american";
            osmt:name "Seamen's Club";
            osmm:type "w";
            osmm:isClosed "true"^^xsd:boolean;
            osmm:loc "Point(53.56017820787392 8.5839141731814)"^^geo:wktLiteral;
            osmm:version "4"^^xsd:integer;
            osmm:user "Lutalica_1974";
            osmm:timestamp "2016-08-17 19:16:40 UTC"^^xsd:dateTime;
            osmm:changeset "41519398"^^xsd:integer.
        "}#",
    },
]
